src/librustc_trait_selection/traits/error_reporting/suggestions.rs

```diff
@@ -1424,96 +1412,122 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
-                // ```
-                debug!("parent_def_kind: {:?}", self.tcx.def_kind(parent_did));
-                let is_raw_borrow_inside_fn_like_call = match self.tcx.def_kind(parent_did) {
-                    DefKind::Fn | DefKind::Ctor(..) => target_ty.is_unsafe_ptr(),
-                    _ => false,
-                };
-
-                if (tables.is_method_call(e) && is_region_borrow)
-                    || is_raw_borrow_inside_fn_like_call
-                {
-                    err.span_help(
-                        parent_span,
-                        "consider moving this into a `let` \
+            if let Some(expr_id) = expr {
+                let expr = hir.expect_expr(expr_id);
+                debug!("target_ty evaluated from {:?}", expr);
+
+                let parent = hir.get_parent_node(expr_id);
+                if let Some(hir::Node::Expr(e)) = hir.find(parent) {
+                    let parent_span = hir.span(parent);
+                    let parent_did = parent.owner.to_def_id();
+                    // ```rust
+                    // impl T {
+                    //     fn foo(&self) -> i32 {}
+                    // }
+                    // T.foo();
+                    // ^^^^^^^ a temporary `&T` created inside this method call due to `&self`
+                    // ```
+                    //
+                    let is_region_borrow =
+                        tables.expr_adjustments(expr).iter().any(|adj| adj.is_region_borrow());
+
+                    // ```rust
+                    // struct Foo(*const u8);
+                    // bar(Foo(std::ptr::null())).await;
+                    //     ^^^^^^^^^^^^^^^^^^^^^ raw-ptr `*T` created inside this struct ctor.
+                    // ```
+                    debug!("parent_def_kind: {:?}", self.tcx.def_kind(parent_did));
+                    let is_raw_borrow_inside_fn_like_call = match self.tcx.def_kind(parent_did) {
+                        DefKind::Fn | DefKind::Ctor(..) => target_ty.is_unsafe_ptr(),
+                        _ => false,
+                    };
+
+                    if (tables.is_method_call(e) && is_region_borrow)
+                        || is_raw_borrow_inside_fn_like_call
+                    {
+                        err.span_help(
+                            parent_span,
+                            "consider moving this into a `let` \
                         binding to create a shorter lived borrow",
-                    );
+                        );
+                    }
                 }
             }
+        } else {
+            if let Some((_, upvar_span)) = upvar {
+                let mut span = MultiSpan::from_span(upvar_span);
+                span.push_span_label(
+                    upvar_span,
+                    format!("has type `{}` which {}", target_ty, trait_explanation),
+                );
+                err.span_note(span, &format!("captured outer value {}", trait_explanation));
+            }
```

Comment:
Most of this is repeated from above, you could do it outside this if/else altogether.

The `span_note` seems like it would work regardless of `interior`, but maybe it's too much info if `interior` also has data. If you want to make that bit conditional, you still can, without duplicating the span label code.